core/src/main/scala/caliban/schema/Schema.scala

```diff
@@ -699,43 +728,66 @@ abstract class PartiallyAppliedFieldBase[V](name: String, description: Option[St
       description,
       _ => Nil,
       () =>
-        if (ev.optional) ev.toType_(ft.isInput, ft.isSubscription)
+        if (ev.optional || ev.canFail) ev.toType_(ft.isInput, ft.isSubscription)
         else ev.toType_(ft.isInput, ft.isSubscription).nonNull,
       isDeprecated = Directives.isDeprecated(directives),
       deprecationReason = Directives.deprecationReason(directives),
-      directives = Some(directives.filter(_.name != "deprecated")).filter(_.nonEmpty)
+      directives = Some(
+        directives.filter(_.name != "deprecated") ++ {
+          if (enableSemanticNonNull && !ev.optional && ev.canFail) Some(Directive("semanticNonNull"))
+          else None
+        }
+      ).filter(_.nonEmpty)
     )
 }
 
-case class PartiallyAppliedField[V](name: String, description: Option[String], directives: List[Directive])
-    extends PartiallyAppliedFieldBase[V](name, description, directives) {
+case class PartiallyAppliedField[V](
+  name: String,
+  description: Option[String],
+  directives: List[Directive],
+  enableSemanticNonNull: Boolean
+) extends PartiallyAppliedFieldBase[V](name, description, directives, enableSemanticNonNull) {
   def either[R, V1](
     fn: V => Either[V1, Step[R]]
   )(implicit ev: Schema[R, V1], ft: FieldAttributes): (__Field, V => Step[R]) =
     (makeField, (v: V) => fn(v).fold(ev.resolve, identity))
 }
 
-case class PartiallyAppliedFieldLazy[V](name: String, description: Option[String], directives: List[Directive])
-    extends PartiallyAppliedFieldBase[V](name, description, directives) {
+case class PartiallyAppliedFieldLazy[V](
+  name: String,
+  description: Option[String],
+  directives: List[Directive],
+  enableSemanticNonNull: Boolean
+) extends PartiallyAppliedFieldBase[V](name, description, directives, enableSemanticNonNull) {
   def either[R, V1](
     fn: V => Either[V1, Step[R]]
   )(implicit ev: Schema[R, V1], ft: FieldAttributes): (__Field, V => Step[R]) =
     (makeField, (v: V) => FunctionStep(_ => fn(v).fold(ev.resolve, identity)))
 }
 
-case class PartiallyAppliedFieldWithArgs[V, A](name: String, description: Option[String], directives: List[Directive]) {
+case class PartiallyAppliedFieldWithArgs[V, A](
+  name: String,
+  description: Option[String],
+  directives: List[Directive],
+  enableSemanticNonNull: Boolean
```

Comment:
I think that since users can provide directives here, we should probably leave that up to them to provide the directive without adding the `enableSemanticNonNull` argument.

Also, maybe we should store the directive somewhere instead of recreating it every time

Reply:
Done!